File: backend/app/api/v1/schemas/booking_request.py
from datetime import datetime
from typing import Any, Optional
from pydantic import BaseModel, ConfigDict, Field


class EnvBookingSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    environment_id: int
    environment_name: Optional[str] = None
    project_name: Optional[str] = None
    start_date: datetime
    end_date: datetime
    status: str
    # REQUIRED SINCE 2026-08-08, and it is worth knowing why it wasn't.
    # Defaulted `False` from the day it shipped and set by NO construction
    # site, this field answered "no conflicts" for every booking on every
    # endpoint that returns this type. Nothing rendered it — the conflict
    # indicators all read `BookingResponse` — so it was dead weight rather than
    # a visible lie, and it stayed that way because a default cannot fail.
    # It is now populated from `conflict_service.bookings_with_unacknowledged_conflicts`,
    # batched once per response, and required so a new site cannot repeat this.
    has_unacknowledged_conflicts: bool
    # A3's usage-agreement warning — see BookingResponse for what it means.
    #
    # REQUIRED, NOT DEFAULTED, deliberately unlike the optional fields above:
    # EnvBookingSummary is constructed by keyword at six sites across two
    # routers, and a default would let a missed one render a booking as "no
    # gap" while `GET /bookings` reports the same booking as in gap. A2 left
    # this exact type self-contradictory that way. Pydantic raises on a missing
    # required field, so a new construction site cannot forget — though note
    # that guards only OMISSION: passing a constant satisfies it just as well,
    # which is why every site has a named test.
    agreement_gap: Optional[str]
    has_unacknowledged_agreement_gap: bool
    # Provenance, not a live link — see Booking.environment_group_id. Null for
    # a hand-picked environment; set for one that arrived via a group.
    environment_group_id: Optional[int] = None
    environment_group_name: Optional[str] = None


class BookingRequestCreate(BaseModel):
    project_name: str
    # The project this booking belongs to. Distinct from `project_name`, which
    # is free text the UI now labels "Purpose" — see the spec.
    project_id: Optional[int] = None
    booking_type_id: int
    start_date: datetime
    end_date: datetime
    # May be empty when environment_group_ids supplies at least one group —
    # the combined "at least one environment" rule is enforced by the
    # service, not by Pydantic, because it spans both fields.
    environment_ids: list[int] = Field(default_factory=list)
    environment_group_ids: list[int] = Field(default_factory=list)
    notes: Optional[str] = None
    context_tag: str = "none"
    exclusive_use_requested: bool = False
    custom_fields: Optional[dict[str, Any]] = None
    delegate_user_ids: Optional[list[int]] = None


class BookingRequestUpdate(BaseModel):
    project_name: Optional[str] = None
    # The project this booking belongs to. Distinct from `project_name`, which
    # is free text the UI now labels "Purpose" — see the spec.
    project_id: Optional[int] = None
    booking_type_id: Optional[int] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    notes: Optional[str] = None
    context_tag: Optional[str] = None
    exclusive_use_requested: Optional[bool] = None
    delegate_user_ids: Optional[list[int]] = None


class BookingRequestCustomFieldsUpdate(BaseModel):
    values: dict[str, Any]


class AddEnvironmentRequest(BaseModel):
    environment_id: int
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None


class BookingRequestResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    tenant_id: int
    project_name: str
    project_id: Optional[int] = None
    project_name_link: Optional[str] = None  # the Project's name, if linked
    booking_type_id: int
    start_date: datetime
    end_date: datetime
    notes: Optional[str]
    context_tag: str
    exclusive_use_requested: bool
    custom_fields: Optional[dict[str, Any]]
    booked_by: int
    delegate_user_ids: Optional[list[int]]
    rollup_status: str
    bookings: list[EnvBookingSummary]


class BookingRequestCreateResponse(BaseModel):
    request: BookingRequestResponse
    detected_conflicts: dict[int, list[EnvBookingSummary]]
    # `booking_id -> message` for the bookings JUST CREATED that no live usage
    # agreement covers. Keyed by booking id, not environment id (unlike
    # detected_conflicts, whose key is the environment being contended for),
    # because the gap is a property of one booking and a group booking may hold
    # several bookings against the same environment over different dates.
    #
    # Absent keys mean "no gap"; an empty map is the ordinary case and is what
    # keeps this a warning rather than a banner. The same text is on each
    # booking's own summary — this map exists so the caller does not have to
    # walk `request.bookings` to find out whether to say anything at all.
    agreement_gaps: dict[int, str] = {}


class PreviewConflictsRequest(BaseModel):
    environment_ids: list[int]
    start_date: datetime
    end_date: datetime


class PreviewConflictsResponse(BaseModel):
    conflicts: dict[int, list[EnvBookingSummary]]

File: frontend/src/types/bookingRequest.ts
export type EnvBookingSummary = {
  id: number;
  environment_id: number;
  environment_name?: string | null;
  project_name?: string | null;
  start_date: string;
  end_date: string;
  status: string;
  has_unacknowledged_conflicts?: boolean;
  // A3's usage-agreement warning — see BookingResponse in booking.ts for
  // what it means. REQUIRED here, not optional/defaulted, deliberately
  // unlike has_unacknowledged_conflicts immediately above: the backend's
  // EnvBookingSummary types this the same way (no default) precisely
  // because it is constructed by keyword at six call sites across two
  // routers, and a default would let a missed one silently render "no gap"
  // for a booking that has one.
  agreement_gap: string | null;
  has_unacknowledged_agreement_gap: boolean;
  // Provenance, not a live link — set for a booking that arrived via an
  // environment group, null for a hand-picked environment.
  environment_group_id: number | null;
  environment_group_name: string | null;
};

export type BookingRequestResponse = {
  id: number;
  tenant_id: number;
  project_name: string;
  // The project this request belongs to. Distinct from `project_name`, which
  // is free text the UI labels "Purpose" — see CLAUDE.md's project_name_link
  // note. `project_name_link` is the linked Project's name, if any.
  project_id: number | null;
  project_name_link: string | null;
  booking_type_id: number;
  start_date: string;
  end_date: string;
  notes: string | null;
  context_tag: string;
  exclusive_use_requested: boolean;
  custom_fields: Record<string, unknown> | null;
  booked_by: number;
  delegate_user_ids: number[] | null;
  rollup_status: string;
  bookings: EnvBookingSummary[];
};

export type BookingRequestCreatePayload = {
  project_name: string;
  // The project this request belongs to. Distinct from `project_name` above.
  project_id?: number | null;
  booking_type_id: number;
  start_date: string;
  end_date: string;
  environment_ids: number[];
  // May be empty when environment_ids supplies at least one environment —
  // the combined "at least one environment" rule is enforced by the
  // backend service, since it spans both fields.
  environment_group_ids?: number[];
  notes?: string | null;
  context_tag?: string;
  exclusive_use_requested?: boolean;
  custom_fields?: Record<string, unknown> | null;
  delegate_user_ids?: number[] | null;
};

export type BookingRequestCreateResponse = {
  request: BookingRequestResponse;
  detected_conflicts: Record<number, EnvBookingSummary[]>;
  // `booking_id -> message` for the bookings JUST CREATED that no live usage
  // agreement covers. Keyed by booking id (not environment id, unlike
  // detected_conflicts) because the gap is a property of one booking and a
  // group booking may hold several bookings against the same environment
  // over different dates. The same text is also on each booking's own
  // `agreement_gap` in `request.bookings` — this map exists so the caller
  // doesn't have to walk that list to know whether to say anything at all.
  // Absent keys mean "no gap"; an empty map is the ordinary case.
  agreement_gaps: Record<number, string>;
};

export type PreviewConflictsResponse = {
  conflicts: Record<number, EnvBookingSummary[]>;
};
